### ISSUE TITLE
feat: expose EnableCheckpointing on ICompiledTrainingPlan interface

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#165]
## Motivation

`CompiledTrainingPlan<T>.EnableCheckpointing(int segmentSize)` exists and works (see `src/AiDotNet.Tensors/Engines/Compilation/CompiledTrainingPlan.cs:100`), but the class is `internal sealed` so consumer code that receives `ICompiledTrainingPlan<T>` from `CompiledModelCache.GetOrCompileTraining` cannot call it without reflection. This blocks AiDotNet from wiring gradient checkpointing into `TrainWithTape` for large-model memory reduction.

## Proposed change

Add `EnableCheckpointing` to the `ICompiledTrainingPlan<T>` interface:

```csharp
public interface ICompiledTrainingPlan<T> : IDisposable
{
    Tensor<T> Step();
    Tensor<T>[] Gradients { get; }
    int ForwardStepCount { get; }
    int BackwardStepCount { get; }
    void ConfigureOptimizer(...);

    /// <summary>
    /// Enables gradient checkpointing for this plan, reducing memory from
    /// O(N) to O(sqrt(N)) at the cost of ~33% more compute (each segment's
    /// forward runs twice during backward).
    /// </summary>
    /// <param name="segmentSize">Steps per checkpoint segment. 0 = auto (sqrt(N)).</param>
    void EnableCheckpointing(int segmentSize = 0);
}
```

The implementation already exists on the concrete class — just surface it.

## Consumer-side wiring after this ships

In `NeuralNetworks/NeuralNetworkBase.TrainWithTape` we'll wire a single `plan.EnableCheckpointing(segmentSize: 0)` call at plan-creation time, gated on a builder-level opt-in (`ConfigureActivationCheckpointing()`). Memory goes from O(layers) to O(sqrt(layers)) with ~33% compute overhead — the right tradeoff for large transformers training on memory-constrained hardware.

## Acceptance criteria

- `EnableCheckpointing(int)` on `ICompiledTrainingPlan<T>` with matching default parameter
- Unit test: `plan.EnableCheckpointing(4)` followed by `plan.Step()` produces identical gradients to non-checkpointed `plan.Step()` within floating-point tolerance
- Memory profiling test showing peak allocation reduction matches the expected sqrt(N) ceiling

## Reference

Plan: `lucky-waddling-knuth` (AiDotNet repo). Consumer PR is blocked on this landing.
